### PR TITLE
[STORM-3600] Add caching in Cluster.calculateSharedOffHeapNodeMemory

### PR DESCRIPTION
**Pull request by bipinprasad**
This yields a 4 fourfold increase in speed when scheduling large cluster. Tested via TestLargeCluster.
Anonymized topologies and configurations generated by TestTopologyAnonymizerUtils.

### CODE REVIEW

**Comment by Ethanlm** [comment]
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.0.0:check (validate) on project storm-server: You have 1 Checkstyle violation. -> [Help 1]
```
There is a checkstyle violation. Can you please fix it? Thanks

File: storm-server/src/main/java/org/apache/storm/scheduler/Cluster.java [hunk 1]
```diff
@@ -709,6 +720,33 @@ public void assign(SchedulerAssignment assignment, boolean ignoreSingleException
...
+     * @param td TopologyDetails to examine
+     */
+    private void initializeTopoSharedOffHeapNodeMemoryFlag(TopologyDetails td) {
+        String topoId = td.getId();
+        topoSharedOffHeapMemoryNodeFlag.put(topoId, false);
+        StormTopology topology = td.getTopology();
+        if (topology == null) {
+            return; // accomodate multitenant_scheduler_test.clj
+        }
+        if (topology.is_set_component_to_shared_memory()) {
```
Comment (Ethanlm) [inline storm-server/src/main/java/org/apache/storm/scheduler/Cluster.java:735]: We should be able to simplify this with 

```
        StormTopology topology = td.getTopology();
        if (topology.is_set_shared_memory()) {
            for (SharedMemory sharedMemory : topology.get_shared_memory().values()) {
                double val = sharedMemory.get_off_heap_node();
                if (val > 0.0) {
                    topoSharedOffHeapMemoryNodeFlag.put(topoId, true);
                    return;
                }
            }
        }
        
```

File: storm-server/src/main/java/org/apache/storm/scheduler/Cluster.java [hunk 2]
```diff
@@ -749,12 +792,16 @@ private double calculateSharedOffHeapNodeMemory(
      */
     public void freeSlot(WorkerSlot slot) {
         // remove the slot from the existing assignments
+        String nodeId = slot.getNodeId();
```
Comment (Ethanlm) [inline storm-server/src/main/java/org/apache/storm/scheduler/Cluster.java:795]: Since this is added, Line 808 can be updated to use `nodeId` instead of `slot.getNodeId()`

File: storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+    public static byte[] getResourceAsBytes(String resource) throws Exception {
+        InputStream in = getResourceAsStream(resource);
+        if (in == null) {
+            return null;
+        }
+        try (ByteArrayOutputStream out = new ByteArrayOutputStream()) {
+            while (in.available() > 0) {
+                out.write(in.read());
+            }
+            return out.toByteArray();
```
Comment (Ethanlm) [inline storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java:139]: I was wondering if `return` within the try-with-resources block will close the resources or not. And I found this https://stackoverflow.com/questions/22947755/try-with-resources-and-return-statements-in-java/22947904

So it should be fine.

File: storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+                Config.TOPOLOGY_PRIORITY,
+                Config.TOPOLOGY_WORKERS,
+                Config.TOPOLOGY_WORKER_MAX_HEAP_SIZE_MB,
+                Config.TOPOLOGY_SUBMITTER_USER,
+                Config.TOPOLOGY_ACKER_CPU_PCORE_PERCENT,
+                Config.TOPOLOGY_ACKER_RESOURCES_OFFHEAP_MEMORY_MB,
+                Config.TOPOLOGY_ACKER_RESOURCES_ONHEAP_MEMORY_MB,
+        };
+
+        for (String nm : codeResourceMap.keySet()) {
```
Comment (Ethanlm) [inline storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java:186]: It looks like `nm` is actually `topoId`? What does `nm` stand for?

File: storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+     *
+     * @param uniformSupervisors true if all supervisors are of the same size, false otherwise.
+     * @return supervisor details indexed by id
+     */
+    private static Map<String, SupervisorDetails> createSupervisors(boolean uniformSupervisors) {
+        Map<String, SupervisorDetails> retVal;
+        if (uniformSupervisors) {
+            int numRacks = 16;
+            int numSupersPerRack = 82;
+            int numPorts = 50; // scheduling will fill up 2-6, leaving of 90% workerslots unused - does this cause slow scheduling?
```
Comment (Ethanlm) [inline storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java:318]: What does the comment mean? Or can it be removed?

File: storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+            double cpu = 7200; // %percent
+            double mem = 356_000; // MB
+            Map<String, Double> miscResources = new HashMap<>();
+            miscResources.put("network.resource.units", 100.0);
+
+            return TestUtilsForResourceAwareScheduler.genSupervisorsWithRacks(
+                    numRacks, numSupersPerRack, numPorts, rackStart, superInRackStart, cpu, mem, miscResources);
+
+        } else {
+            // this non-uniform supervisor distribution closelt (but not exactly) mimics IridiumBlue cluster
```
Comment (Ethanlm) [inline storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java:330]: `IridiumBlue` doesn't mean anything in the community code. Maybe reword it.

Typo: `closelt`

File: storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java [hunk 7]
```diff
@@ -0,0 +1,458 @@
...
+
+        for (TopologyDetails td : topoDetailsArray) {
+            TestUtilsForResourceAwareScheduler.assertTopologiesFullyScheduled(cluster, td.getName());
+        }
+
+        // Remove topology and reschedule it
+        java.util.Random random = new java.util.Random();
+        for (int i = 0 ; i < topoDetailsArray.length ; i++) {
+            startTime = System.currentTimeMillis();
+            int iTopo = i; // random.nextInt(topoDetailsArray.length);
```
Comment (Ethanlm) [inline storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java:406]: Which one we want to keep? ` iTopo = i` or `random.nextInt(topoDetailsArray.length);`

File: storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java [hunk 8]
```diff
@@ -0,0 +1,458 @@
...
+
+        INimbus iNimbus = new INimbusTest();
+        Cluster cluster = new Cluster(iNimbus, new ResourceMetrics(new StormMetricsRegistry()), supervisors, new HashMap<>(),
+                topologies, confWithDefaultStrategy);
+
+        scheduler = new ResourceAwareScheduler();
+
+        List<Class> classesToDebug = Arrays.asList(DefaultResourceAwareStrategy.class,
+                GenericResourceAwareStrategy.class, ResourceAwareScheduler.class,
+                Cluster.class // count calls to calculateSharedOffHeapNodeMemory()
```
Comment (Ethanlm) [inline storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java:388]: What does the comment mean here?

File: storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestTopologyAnonymizerUtils.java [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+
+    private String originalResourcePath;
+    private String outputDirPath;
+
+    public TestTopologyAnonymizerUtils() {
+        this.originalResourcePath = DEFAULT_ORIGINAL_RESOURCES_PATH;
+        this.outputDirPath = DEFAULT_ANONYMIZED_RESOURCES_OUTDIR;
+    }
+
+    //public TestTopologyAnonymizerUtils(String originalResourcePath, String outputDirPath) {
```
Comment (Ethanlm) [inline storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestTopologyAnonymizerUtils.java:79]: Maybe clean up the comments if they are not needed

File: storm-server/src/main/java/org/apache/storm/scheduler/Cluster.java (hunk 1 above)
Comment (bipinprasad) [reply storm-server/src/main/java/org/apache/storm/scheduler/Cluster.java:735]: (1) topology == null check is required weirdly enough due to some other tests failure (later I also found note in TopologyDetails.getSharedMemoryRequests()
(2) In StormTopology shared_memory and component_to_shared_memory are two different thrift fields. And setters don't seem to cross set. If I change this code then this will change the prior logic in Cluster.java. How are these two different fields equivalent?

File: storm-server/src/main/java/org/apache/storm/scheduler/Cluster.java (hunk 2 above)
Comment (bipinprasad) [reply storm-server/src/main/java/org/apache/storm/scheduler/Cluster.java:795]: changed

File: storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestTopologyAnonymizerUtils.java (hunk 9 above)
Comment (bipinprasad) [reply storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestTopologyAnonymizerUtils.java:79]: removed

File: storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java (hunk 4 above)
Comment (bipinprasad) [reply storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java:186]: nm (name) is the crud that stands in from of the code.ser and conf.ser in the file name and is the same for two topology and config files for the same topology. The code is expecting/assuming topology name to be embedded in the file names.

File: storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java (hunk 5 above)
Comment (bipinprasad) [reply storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java:318]: I found that for topology components with large cpu/mem, most of the workeslots get left unused - but the looping logic in scheduling spends a lot of cycles in those. How workerslots are processed may need to be revisited for further speed gains. If it is not clear, it needs to be reworded.

File: storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java (hunk 6 above)
Comment (bipinprasad) [reply storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java:330]: changed

File: storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java (hunk 8 above)
Comment (bipinprasad) [reply storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java:388]: I had debug statement earlier in Cluster.java - since removed. Comment is not relevant any more.
Removed.

File: storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java (hunk 7 above)
Comment (bipinprasad) [reply storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java:406]: i for now.

File: storm-server/src/main/java/org/apache/storm/scheduler/Cluster.java (hunk 1 above)
Comment (Ethanlm) [reply storm-server/src/main/java/org/apache/storm/scheduler/Cluster.java:735]: I agree with (1). Sorry I didn't copy that part of the code. 

(2) 
These two fields are equivalent. 
There is a 1-to-1 mapping between `SharedMemory name` to `SharedMemory object`.

shared_memory and component_to_shared_memory are different structures to keep `SharedMemory object` but they essentially keep the whole mapping of `SharedMemory name` to `SharedMemory object`
https://github.com/apache/storm/blob/master/storm-client/src/jvm/org/apache/storm/topology/TopologyBuilder.java#L631-L638

Comment (Ethanlm) [reply storm-server/src/main/java/org/apache/storm/scheduler/Cluster.java:735]: They are designed this way. For example

I can have two SharedMemory with name as `sharedMemory1` and `sharedMemory2`. 

And I can add  `sharedMemory1` to `comp1` and `sharedMemory2` to `comp2`.

Now shared_memory is 
 object1 (name=`sharedMemory1`),  object2 (name=`sharedMemory2`);

component_to_shared_memory is 

`comp1` -> object1 (name=`sharedMemory1`), `comp2` -> object2 (name=`sharedMemory2`);

File: storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java (hunk 4 above)
Comment (Ethanlm) [reply storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java:186]: Thanks.  I was confused about `nm` variable. Maybe `name` is more clear. But that's fine.

File: storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java (hunk 5 above)
Comment (Ethanlm) [reply storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java:318]: It would be better if it can be reworded. Your comment here is much clearer to me

File: storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java (hunk 7 above)
Comment (Ethanlm) [reply storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java:406]: Maybe clean it up to avoid confusion. Pick one

File: storm-server/src/main/java/org/apache/storm/scheduler/Cluster.java (hunk 1 above)
Comment (bipinprasad) [reply storm-server/src/main/java/org/apache/storm/scheduler/Cluster.java:735]: Thanks for explanation. Looks like TopologyBuilder.createTopology() makes the only to call to set_component_to_shared_memory() and set_shared_memory() and thus ensuring this consistency.

Changed code.

File: storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java (hunk 7 above)
Comment (bipinprasad) [reply storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java:406]: changed

File: storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java (hunk 3 above)
Comment (bipinprasad) [reply storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java:139]: Yes. Pretty cool JDK 8 feature. Here is more authoritative info - which actually uses return as an example too: https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html

File: storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java [hunk 10]
```diff
@@ -0,0 +1,456 @@
...
+        List<String> fileNames = new ArrayList<>();
+
+        try (
+                InputStream in = getResourceAsStream(path);
+                BufferedReader br = new BufferedReader(new InputStreamReader(in))
+        ) {
+            String resource;
+
+            while ((resource = br.readLine()) != null) {
+                if (resource.endsWith("code.ser") || resource.endsWith("conf.ser")) {
```
Comment (govind-menon) [inline storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java:103]: Nit: move these file extensions as constants to the head of the file

File: storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java (hunk 3 above)
Comment (govind-menon) [reply storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java:139]: Agreed - try with resources will close as part of the "destructor"

File: storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java [hunk 11]
```diff
@@ -0,0 +1,456 @@
...
+        List<String> errors = new ArrayList<>();
+        List<String> resources = getResourceFiles(TEST_RESOURCE_PATH);
+        Map<String, String> codeResourceMap = new TreeMap<>();
+        Map<String, String> confResourceMap = new HashMap<>();
+        for (int i = 0 ; i < resources.size() ; i++) {
+            String resource = resources.get(i);
+            int idxOfSlash = resource.lastIndexOf("/");
+            int idxOfDash = resource.lastIndexOf("-");
+            String nm = idxOfDash > idxOfSlash ? resource.substring(idxOfSlash + 1, idxOfDash) : resource.substring(idxOfSlash + 1, resource.length() - 7);
+            if (resource.endsWith("code.ser")) {
```
Comment (govind-menon) [inline storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java:166]: Nit: See above about extracting file extension as cosnt ants

Comment (bipinprasad) [reply storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java:166]: changed

File: storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java (hunk 10 above)
Comment (bipinprasad) [reply storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java:103]: changed

File: storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java [hunk 12]
```diff
@@ -70,6 +70,8 @@
 
     public static final String TEST_CLUSTER_NAME = "largeCluster01";
     public static final String TEST_RESOURCE_PATH = "clusterconf/" + TEST_CLUSTER_NAME;
+    public static final String COMPRESSED_SERIALIZED_TOPOLOGY_FILENAME_ENDING = "code.ser";
```
Comment (Ethanlm) [inline storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java:73]: There is a variable with the same name in TestTopologyAnonymizerUtils.java but with different meanings. What's the difference here?

Comment (bipinprasad) [reply storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java:73]: Anonymizer can be stricter with naming convention.  Files in the resources need just the bare minimum to be distinguished from each other. But more importantly, they are independent of each other.
